### PR TITLE
ARROW-1021: [Python] Add documentation for C++ pyarrow API

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
This required a bit more churn than I expected (especially to get the Cython example and test to work). I think this is ready for review now.

**Comment by pitrou** [comment]
Hmm, there's still an AppVeyor failure. Will try to fix :-/

**Comment by pitrou** [comment]
Ok, I'm afraid I don't know how to get the Windows Cython test to work. Here is the log:
```
pyarrow_cython_example.obj : error LNK2001: unresolved external symbol "__declspec(dllimport) public: __int64 __cdecl arrow::Array::length(void)const " (__imp_?length@Array@arrow@@QEBA_JXZ)
C:\Users\appveyor\AppData\Local\Temp\1\pytest-of-appveyor\pytest-0\test_cython_api0\pyarrow_cython_example.cp35-win_amd64.pyd : fatal error LNK1120: 1 unresolved externals
```
(from https://ci.appveyor.com/project/pitrou/arrow/build/1.0.60/job/aruj4pno67s4xpcf#L6242 )

**Comment by wesm** [comment]
I can take a look at the Windows issue (I have a machine to test on)

**Comment by xhochy** [comment]
@pitrou @wesm Note that there are free Windows VMs for 90 days to test on https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/ If there is ever a dark and lonely day, I might try to get my head around this to write an ansible/vagrant deployment that sets up the Vm including a Windows environment as I have never been able to test it myself. But don't expect this to happen soon.

I have not read fully through the license for these VMs but as far as I understand they are made exactly for testing software for Windows.

**Review by xhochy — APPROVED** [review]
This look really great. Can't wait to hand it out to colleagues.

**Comment by pitrou** [comment]
@xhochy I do have a Windows VM at home. But that won't allow me fixing the error, as that requires platform-specific knowledge about the Windows linker that I don't have ;-)

**Comment by wesm** [comment]
Taking a look at this now-ish

**Comment by wesm** [comment]
So I believe the issue on Windows is that we need the `.lib` files in addition to the `.dll` files. It doesn't appear that these are being installed in `%INSTALL_PREFIX%\lib`. Let me tinker with it and see if I can get things to work locally

**Comment by wesm** [comment]
Correction: we need to add `arrow_python` to `libraries` and make sure the `.lib` files (which are being installed) are visible in `library_dirs` in setuptools. I tried to look how Cython deals with NumPy for linking on Windows but it wasn't obvious what's going on. We might need to add a function that will try to locate `arrow_python.lib`, I will try some things to try to at least get the test case working

**Comment by wesm** [comment]
I got this to work locally on Windows -- the documentation will need to be updated to account for the new `get_libraries` and `get_library_dirs` functions. Should we create a `pyarrow.distutils` module or something or leave these at the top level (either is OK by me)

We should try out the example extension with both conda-style and pip-style installations -- I left a TODO because I'm not sure we need to include the `%PYTHON_DIR%\Library\lib` directory in the `library_dirs`

**Comment by pitrou** [comment]
>  Should we create a pyarrow.distutils module or something or leave these at the top level (either is OK by me)

Leaving them at the top-level sounds fine to me.

**Comment by wesm** [comment]
OK, going to put the finishing touches on this now and get the Linux tests passing

File: python/doc/source/api.rst [hunk 1]
```diff
@@ -342,3 +342,15 @@ Apache Parquet
...
+
+.. currentmodule:: pyarrow
+
+Using with C extensions
+-----------------------
+
+.. autosummary::
+   :toctree: generated/
+
+   get_include
```
Comment (xhochy) [inline python/doc/source/api.rst]: Already made me a reminder to use them in `turbodbc` https://github.com/blue-yonder/turbodbc/issues/147. Much appreciated functionality.

**Review by xhochy — APPROVED** [review]

File: python/doc/source/api.rst (hunk 1 above)
Comment (wesm) [reply python/doc/source/api.rst]: We should validate that they work as intended in both conda and pip based installs, I haven't had time to rigorously check

**Comment by wesm** [comment]
Appveyor build running here: https://ci.appveyor.com/project/pitrou/arrow/build/1.0.67. Will merge once that's looking good

**Pull request closed without merging**